qemu/tests/win_driver_base.py

```diff
@@ -0,0 +1,71 @@
+import logging
+
+from virttest import utils_test
+from virttest import error_context
+from qemu.tests import single_driver_install
+
+
+@error_context.context_aware
+def run(test, params, env):
+    """
+    Virtio driver test for windows guest.
+    1) boot guest with virtio device.
+    2) enable and check driver verifier in guest.
+    3) Uninstall and install driver:
+    3.1) uninstall driver.
+    3.2) install driver.
+    3) Downgrade and upgrade driver:
+    3.1) downgrade virtio driver to specified version.
+    3.2) run subtest. (optional)
+    3.3) upgrade virtio driver to original version.
+    4) clear the driver verifier.
+
+    :param test: QEMU test object
+    :param params: Dictionary with the test parameters
+    :param env: Dictionary with test environment
+    """
+    driver = params["driver_name"]
+    timeout = int(params.get("login_timeout", 360))
+    sub_test = params.get("sub_test")
+    cdrom_virtio_downgrade = params.get("cdrom_virtio_downgrade")
+
+    vm = env.get_vm(params["main_vm"])
+    vm.verify_alive()
+    error_context.context("Boot guest with %s device" % driver, logging.info)
+    session = vm.wait_for_login(timeout=timeout)
+    if (params.get("need_enable_verifier"), "yes") == "yes":
+        error_context.context("Enable %s driver verifier in guest" % driver,
+                              logging.info)
+        session = utils_test.qemu.setup_win_driver_verifier(session,
+                                                            driver, vm, timeout)
+
+    try:
+        if params.get("need_uninstall") == "yes":
```

Comment:
1. For uninstall driver, there is no need to do subtest.
2. For downgrade driver, the vm will be re-created with cdrom_virtio_downgrade.

For case single_driver_update, it will only do the operations according to the params.